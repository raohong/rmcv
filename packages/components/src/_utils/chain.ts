const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
