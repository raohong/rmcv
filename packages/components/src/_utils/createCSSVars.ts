import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
