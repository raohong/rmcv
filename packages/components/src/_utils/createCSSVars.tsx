import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
