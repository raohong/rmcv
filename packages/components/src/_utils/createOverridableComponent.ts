import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
