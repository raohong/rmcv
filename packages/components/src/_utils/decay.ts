import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
