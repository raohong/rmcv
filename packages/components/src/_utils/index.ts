export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
