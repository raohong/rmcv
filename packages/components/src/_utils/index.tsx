export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
