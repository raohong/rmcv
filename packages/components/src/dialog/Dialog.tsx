import { useUnmountedRef } from '@rmc-vant/hooks';
import { isEmpty, isString } from '@rmc-vant/utils';
import classNames from 'classnames';
import React, { useState } from 'react';
import { getDataOrAriaProps } from '../_utils';
import Button from '../button';
import { useConfigContext } from '../config-provider';
import Popup from '../popup';
import type { DialogAction, DialogProps } from './interface';

const SMALL_SCREEN_WIDTH = 320;

const Dialog: React.FC<DialogProps> = ({
  title,
  message,
  width,
  className,
  overlay,
  overlayClassName,
  overlayClosable,
  overlayStyle,
  style,
  showCancelButton,
  cancelButtonColor,
  cancelButtonText,
  confirmButtonColor,
  beforeClose,
  onCancel,
  onClose,
  onConfirm,
  lockScroll,
  teleport,
  afterClose,
  visible,
  motionName,
  footer,
  closeOnPopstate = true,
  showConfirmButton = true,
  confirmButtonText,
  messageAlign = 'center',
  theme = 'default',
  ...rest
}) => {
  const { getPrefixCls } = useConfigContext();
  const [loadingState, setLoadingState] = useState<Record<DialogAction, boolean>>({
    cancel: false,
    confirm: false,
  });
  const unmountedRef = useUnmountedRef();

  const handleClose = async (action?: DialogAction) => {
    if (beforeClose && action) {
      if (loadingState[action]) {
        return;
      }
      setLoadingState((prev) => ({
        ...prev,
        [action]: true,
      }));

      try {
        const result = await beforeClose(action);

        if (result === false) {
          return;
        }

        onClose?.();
      } finally {
        if (!unmountedRef.current) {
          setLoadingState((prev) => ({
            ...prev,
            [action]: false,
          }));
        }
      }
    } else {
      onClose?.();
    }
  };

  const handleConfirm = async () => {
    onConfirm?.();
    handleClose('confirm');
  };

  const handleCancel = async () => {
    onCancel?.();
    handleClose('cancel');
  };

  const cls = getPrefixCls('dialog');
  const internalConfirmButtonText = confirmButtonText ?? '确定';
  const internalCancelButtonText = cancelButtonText ?? '取消';
  const hasHeader = !isEmpty(title);
  const hasMessage = !isEmpty(message);
  const hasFooter = !isEmpty(footer) || showCancelButton || showConfirmButton;
  const headerIsIsolated = !hasMessage && !hasFooter;
  const internalMotionName = motionName || getPrefixCls('dialog-animation');
  const footerBorder =
    !isEmpty(footer) ||
    (showCancelButton && showConfirmButton && theme !== 'round-button');
  const buttonBorder = !footerBorder && theme !== 'round-button';
  const titleString = (isString(title) && title) || undefined;
  // const messageString = isString(message) && message;

  const renderFooter = () => {
    if (!isEmpty(footer)) {
      return (
        <div
          className={classNames(
            `${cls}-footer`,
            `${cls}-footer-custom`,
            footerBorder && `${cls}-footer-border`,
          )}
        >
          {footer}
        </div>
      );
    }

    if (!hasFooter) {
      return null;
    }

    return (
      <div
        className={classNames(
          `${cls}-footer`,
          footerBorder && `${cls}-footer-border`,
        )}
      >
        {showCancelButton && (
          <Button
            style={{ color: cancelButtonColor }}
            className={classNames(
              `${cls}-button`,
              `${cls}-button-cancel`,
              showConfirmButton && `${cls}-button-first`,
            )}
            onClick={handleCancel}
            border={buttonBorder}
            loading={loadingState.cancel}
            block
            hairline
          >
            {internalCancelButtonText}
          </Button>
        )}
        {showConfirmButton && (
          <Button
            style={{
              color: confirmButtonColor,
            }}
            onClick={handleConfirm}
            className={classNames(
              `${cls}-button`,
              `${cls}-button-confirm`,
              showCancelButton && `${cls}-button-second`,
            )}
            loading={loadingState.confirm}
            border={buttonBorder}
            block
            hairline
          >
            {internalConfirmButtonText}
          </Button>
        )}
      </div>
    );
  };

  return (
    <Popup
      teleport={teleport}
      lockScroll={lockScroll}
      overlay={overlay}
      overlayClassName={overlayClassName}
      overlayStyle={overlayStyle}
      overlayClosable={overlayClosable}
      onClose={handleClose}
      position="center"
      afterClose={afterClose}
      visible={visible}
      motionName={internalMotionName}
      closeOnPopstate={closeOnPopstate}
      style={{
        width: width === SMALL_SCREEN_WIDTH ? undefined : width,
        ...style,
      }}
      className={classNames(
        cls,
        {
          [`${cls}-theme-round-button`]: theme === 'round-button',
        },
        className,
      )}
      tabIndex={-1}
      role="dialog"
      aria-labelledby={titleString}
      transitionAppear
      aria-modal
      {...getDataOrAriaProps(rest)}
    >
      {hasHeader && (
        <h2
          className={classNames(
            `${cls}-header`,
            headerIsIsolated && `${cls}-header-isolated`,
          )}
        >
          {title}
        </h2>
      )}

      {hasMessage && (
        <div
          className={classNames(`${cls}-message`, {
            [`${cls}-message-align-${messageAlign}`]: messageAlign !== 'center',
            [`${cls}-has-title-message`]: hasHeader,
          })}
        >
          {message}
        </div>
      )}
      {renderFooter()}
    </Popup>
  );
};

export default Dialog;
