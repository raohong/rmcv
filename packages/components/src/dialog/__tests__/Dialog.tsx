import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { getPrefixCls } from '../../_utils';
import Dialog from '../Dialog';

const id = 'dialog';

test('render correctly', () => {
  const tree = render(<Dialog />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with title', () => {
  render(<Dialog title={<span data-testid="title" />} />);

  expect(screen.getByTestId('title')).toBeInTheDocument();
});

test('render with width', () => {
  render(<Dialog width={200} data-testid={id} />);

  expect(screen.getByTestId(id)).toHaveStyle({
    width: '200px',
  });
});

test('render with message', () => {
  render(<Dialog message={<div data-testid="message" />} />);

  expect(screen.getByTestId('message')).toBeInTheDocument();
});

test('render with footer', () => {
  render(<Dialog data-testid={id} footer={<div data-testid="footer" />} />);

  expect(screen.getByTestId('footer')).toBeInTheDocument();
  expect(
    screen.getByTestId(id).querySelector(`.${getPrefixCls('dialog-footer')}`),
  ).toHaveClass(getPrefixCls('dialog-footer-custom'));
});

test('render with theme', () => {
  render(<Dialog theme="round-button" data-testid={id} />);

  expect(screen.getByTestId(id)).toHaveClass(
    getPrefixCls('dialog-theme-round-button'),
  );
});

test('render with showConfirmButton', () => {
  render(<Dialog showConfirmButton={false} data-testid={id} />);

  expect(
    screen
      .getByTestId(id)
      .querySelector(`.${getPrefixCls('dialog-button-confirm')}`),
  ).toBeNull();
});

test('render with showCancelButton', () => {
  render(<Dialog showCancelButton data-testid={id} />);

  expect(
    screen.getByTestId(id).querySelector(`.${getPrefixCls('dialog-button-cancel')}`),
  ).not.toBeNull();
});

test('render with confirmButtonText', () => {
  render(<Dialog confirmButtonText="button-text" data-testid={id} />);

  expect(screen.getByText('button-text')).not.toBeNull();
});

test('render with confirmButtonColor', () => {
  render(<Dialog confirmButtonColor="red" data-testid={id} />);

  expect(
    screen
      .getByTestId(id)
      .querySelector(`.${getPrefixCls('dialog-button-confirm')}`),
  ).toHaveStyle({
    color: 'red',
  });
});

test('render with cancelButtonText', () => {
  render(
    <Dialog showCancelButton cancelButtonText="button-text" data-testid={id} />,
  );

  expect(screen.getByText('button-text')).not.toBeNull();
});

test('render with cancelButtonColor', () => {
  render(<Dialog showCancelButton cancelButtonColor="blue" data-testid={id} />);

  expect(
    screen.getByTestId(id).querySelector(`.${getPrefixCls('dialog-button-cancel')}`),
  ).toHaveStyle({
    color: 'blue',
  });
});

test('render with visible', () => {
  const com = render(<Dialog visible data-testid={id} />);

  expect(screen.getByTestId(id)).toHaveAttribute('aria-hidden', 'false');
});

test('render with onConfirm', () => {
  const onConfirm = jest.fn();

  render(<Dialog visible onConfirm={onConfirm} data-testid={id} />);

  userEvent.click(
    screen
      .getByTestId(id)
      .querySelector(`.${getPrefixCls('dialog-button-confirm')}`)!,
  );

  expect(onConfirm).toHaveBeenCalled();
});
