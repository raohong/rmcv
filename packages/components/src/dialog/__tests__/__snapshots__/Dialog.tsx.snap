// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="rmcv-overlay"
    style="z-index: 1001; display: none;"
  />
  <div
    aria-hidden="true"
    aria-modal="true"
    class="rmcv-popup rmcv-popup-center rmcv-popup-round rmcv-dialog"
    role="dialog"
    style="display: none; z-index: 1002;"
    tabindex="-1"
  >
    <div
      class="rmcv-dialog-footer"
    >
      <button
        class="rmcv-button rmcv-button-hairline rmcv-button-block rmcv-button-size-normal rmcv-dialog-button rmcv-dialog-button-confirm rmcv-touchable"
        role="button"
        style="cursor: cursor;"
      >
        <span
          class="rmcv-button-content"
        >
          确定
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
